flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java

```diff
@@ -155,38 +129,66 @@ public void shutdown() {
 			.set(HttpHeaders.Names.HOST, configuredTargetAddress + ":" + configuredTargetPort)
 			.set(HttpHeaders.Names.CONNECTION, HttpHeaders.Values.CLOSE);
 
-		synchronized (this) {
-			// This ensures strict sequential processing of requests.
-			// If we send new requests immediately we can no longer make assumptions about the order in which responses
-			// arrive, due to which the handler cannot know which future he should complete (not to mention what response
-			// type to read).
-			CompletableFuture<P> nextFuture = lastFuture
-				.handleAsync((f, e) -> submitRequest(httpRequest, messageHeaders), directExecutor)
-				.thenCompose((future) -> future);
-
-			lastFuture = nextFuture;
-			return nextFuture;
-		}
+		return submitRequest(httpRequest, messageHeaders);
 	}
 
 	private <M extends MessageHeaders<R, P, U>, U extends ParameterMapper, R extends RequestBody, P extends ResponseBody> CompletableFuture<P> submitRequest(FullHttpRequest httpRequest, M messageHeaders) {
-		CompletableFuture<P> responseFuture = handler.expectResponse(messageHeaders.getResponseClass());
-
-		try {
-			// write request
-			Channel channel = bootstrap.connect(configuredTargetAddress, configuredTargetPort).sync().channel();
-			channel.writeAndFlush(httpRequest);
-			channel.closeFuture();
-		} catch (InterruptedException e) {
-			return FutureUtils.completedExceptionally(e);
+		synchronized (lock) {
+			CompletableFuture<P> responseFuture = ClientHandler.addHandlerForResponse(bootstrap, sslEngine, messageHeaders.getResponseClass());
```

Comment:
Can't we do it by having a fixed channel initializer which always creates a new `ClientHandler`? Then we can retrieve the actual `ClientHandler` by `channel.pipeline.get(ClientHandler.class)`. And then we can gain access to the future.